File: sections/14_core_patterns_mixed_contest/problems/v_smallest_separated_rearrangement/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string s, answer;
    cin >> s;
    array<int, 26> count{};
    for (char c : s) ++count[c - 'a'];
    int previous = -1;
    for (int position = 0; position < (int)s.size(); ++position) {
        array<int, 27> prefix{}, suffix{};
        for (int c = 0; c < 26; ++c) {
            prefix[c + 1] = max(prefix[c], count[c]);
        }
        for (int c = 25; c >= 0; --c) {
            suffix[c] = max(suffix[c + 1], count[c]);
        }
        int remaining = (int)s.size() - position - 1;
        int chosen = -1;
        for (int c = 0; c < 26; ++c) {
            if (c == previous || count[c] == 0) continue;
            int same = count[c] - 1;
            int other = max(prefix[c], suffix[c + 1]);
            if (same <= remaining / 2
                    && other <= (remaining + 1) / 2) {
                chosen = c;
                break;
            }
        }
        if (chosen == -1) {
            cout << "IMPOSSIBLE\n";
            return 0;
        }
        answer.push_back(char('a' + chosen));
        --count[chosen];
        previous = chosen;
    }
    cout << answer << '\n';
}

File: sections/14_core_patterns_mixed_contest/problems/v_smallest_separated_rearrangement/solution.py
import sys


def main() -> None:
    s = sys.stdin.buffer.readline().strip().decode()
    count = [0] * 26
    for char in s:
        count[ord(char) - ord("a")] += 1
    answer = []
    previous = -1
    for position in range(len(s)):
        prefix = [0] * 27
        suffix = [0] * 27
        for char in range(26):
            prefix[char + 1] = max(prefix[char], count[char])
        for char in range(25, -1, -1):
            suffix[char] = max(suffix[char + 1], count[char])
        remaining = len(s) - position - 1
        chosen = -1
        for char in range(26):
            if char == previous or count[char] == 0:
                continue
            same = count[char] - 1
            other = max(prefix[char], suffix[char + 1])
            if same <= remaining // 2 and other <= (remaining + 1) // 2:
                chosen = char
                break
        if chosen == -1:
            print("IMPOSSIBLE")
            return
        answer.append(chr(ord("a") + chosen))
        count[chosen] -= 1
        previous = chosen
    print("".join(answer))


if __name__ == "__main__":
    main()
